lib/PuppeteerSharp/ExecutionContext.cs

```diff
@@ -22,44 +24,86 @@ public ExecutionContext(Session client, ContextPayload contextPayload, Func<dyna
         }
 
-        internal async Task<JSHandle> EvaluateHandleAsync(string pageFunction, object[] args)
+        internal async Task<JSHandle> EvaluateHandleAsync(string script, object[] args)
         {
-            if (!string.IsNullOrEmpty(pageFunction))
+            if (string.IsNullOrEmpty(script))
             {
-                dynamic remoteObject;
+                return null;
+            }
+
+            if (IsFunction(script))
+            {
+                return await EvaluateFunctionAsync(script, args);
+            }
+
+            return await EvaluateExpressionAsync(script);
+
+        }
 
-                try
+        private async Task<JSHandle> EvaluateExpressionAsync(string script)
+        {
+            dynamic remoteObject;
+
+            try
+            {
+                remoteObject = await _client.SendAsync("Runtime.evaluate", new Dictionary<string, object>()
                 {
-                    remoteObject = await _client.SendAsync("Runtime.evaluate", new Dictionary<string, object>()
-                    {
-                        ["expression"] = pageFunction,
-                        ["contextId"] = _contextId,
-                        ["returnByValue"] = false,
-                        ["awaitPromise"] = true
-                    });
-
-                    return ObjectHandleFactory(remoteObject.result);
-                }
-                catch (Exception ex)
+                    {"contextId", _contextId},
+                    {"expression", script},
+                    {"returnByValue", false},
+                    {"awaitPromise", true}
+                });
+
+                return ObjectHandleFactory(remoteObject.result);
+            }
+            catch (Exception ex)
+            {
+                throw new EvaluationFailedException("Evaluation Failed", ex);
+            }
+        }
+
+        private async Task<JSHandle> EvaluateFunctionAsync(string script, object[] args)
+        {
+            dynamic result = await _client.SendAsync("Runtime.callFunctionOn", new Dictionary<string, object>()
                 {
-                    throw new EvaluationFailedException("Evaluation Failed", ex);
-                }
+                    {"functionDeclaration", script },
+                    {"executionContextId", _contextId},
+                    {"arguments", FormatArguments(args)},
+                    {"returnByValue", false},
+                    {"awaitPromise", true}
+                });
+
+            if (result.exceptionDetails != null)
+            {
+                throw new EvaluationFailedException("Evaluation failed: " +
+                    Helper.GetExceptionMessage(result.exceptionDetails.ToObject<EvaluateExceptionDetails>()));
             }
 
-            return null;
+            return ObjectHandleFactory(result.result);
+        }
+
+        private object FormatArguments(object[] args)
+        {
+            return args.Select(o => new { value = o });
```

Comment:
this is not the complete implementation of the [`convertArgument`](https://github.com/GoogleChrome/puppeteer/blob/v1.0.0/lib/ExecutionContext.js#L78) method

Reply:
I would wait for test needing that code. What do you think?

Reply:
ok, I wasn't sure if this method has tests